src/mds/Server.cc

```diff
@@ -5861,6 +5876,34 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur)
     auto pi = cur->project_inode(mdr);
     cur->setxattr_ephemeral_dist(val);
     pip = pi.inode.get();
+  } else if (name == "ceph.dir.rctime") {
+    if (!cur->is_dir()) {
+      respond_to_request(mdr, -EINVAL);
+      return;
+    }
+
+    uint64_t sec, nsec = 0;
+    try {
+      if (auto pos = value.find('.', 0); pos != string::npos) {
+        sec  = boost::lexical_cast<uint64_t>(value.substr(0, pos));
+        nsec = boost::lexical_cast<uint64_t>(value.substr(pos+1));
+      } else {
+        sec = boost::lexical_cast<uint64_t>(value);
+      }
+    } catch (boost::bad_lexical_cast const&) {
+      dout(10) << "bad vxattr value, unable to parse utime_t for " << name << dendl;
+      respond_to_request(mdr, -EINVAL);
+      return;
+    }
+    utime_t val(sec, nsec);
+
+    if (!xlock_nestlock(mdr, cur))
```

Comment:
Fails with the following crash here: https://pulpito.ceph.com/vshankar-2022-02-05_17:27:49-fs-wip-vshankar-testing-20220201-113815-testing-default-smithi/6663067/

```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-10531-g09fc2e9e/rpm/el8/BUILD/ceph-17.0.0-10531-g09fc2e9e/src/mds/Locker.cc: 4782: ceph_abort_msg("abort() called")

 ceph version 17.0.0-10531-g09fc2e9e (09fc2e9e470435f1e6dbf462a51853391f4a4cc0) quincy (dev)
 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xe5) [0x7f83dd2ec91f]
 2: (Locker::simple_lock(SimpleLock*, bool*)+0x588) [0x560d6e1a4108]
 3: (Locker::xlock_start(SimpleLock*, boost::intrusive_ptr<MDRequestImpl>&)+0x380) [0x560d6e1a5dd0]
 4: (Locker::acquire_locks(boost::intrusive_ptr<MDRequestImpl>&, MutationImpl::LockOpVec&, CInode*, bool)+0x24a0) [0x560d6e1bb8c0]
 5: (Server::xlock_nestlock(boost::intrusive_ptr<MDRequestImpl>&, CInode*)+0xd2) [0x560d6e00fab2]
 6: (Server::handle_set_vxattr(boost::intrusive_ptr<MDRequestImpl>&, CInode*)+0x14d4) [0x560d6e036364]
 7: (Server::handle_client_setxattr(boost::intrusive_ptr<MDRequestImpl>&)+0xcd) [0x560d6e03979d]
 8: (Server::dispatch_client_request(boost::intrusive_ptr<MDRequestImpl>&)+0xc7b) [0x560d6e0600ab]
 9: (Server::handle_client_request(boost::intrusive_ptr<MClientRequest const> const&)+0x566) [0x560d6e06bda6]
 10: (Server::dispatch(boost::intrusive_ptr<Message const> const&)+0x133) [0x560d6e074fe3]
 11: (MDSRank::handle_message(boost::intrusive_ptr<Message const> const&)+0xd3c) [0x560d6dfcb74c]
 12: (MDSRank::_dispatch(boost::intrusive_ptr<Message const> const&, bool)+0x7cb) [0x560d6dfce39b]
 13: (MDSRankDispatcher::ms_dispatch(boost::intrusive_ptr<Message const> const&)+0x5c) [0x560d6dfce9bc]
 14: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x108) [0x560d6dfbd8b8]
 15: (DispatchQueue::entry()+0x14fa) [0x7f83dd57240a]
 16: (DispatchQueue::DispatchThread::entry()+0x11) [0x7f83dd628891]
 17: /lib64/libpthread.so.0(+0x814a) [0x7f83dc2c614a]
 18: clone()
```

I think this needs to be a wrlock rather than an xlock.

Also, BTW, as per comment here(*), seems that we may want to acquire some additional locks? @batrick @gregsfortytwo What is dirlock? Seems to me like it is something existed and is got removed?

* https://github.com/ceph/ceph/blob/master/src/mds/MDCache.cc#L2105

Reply:
> Also, BTW, as per comment here(*), seems that we may want to acquire some additional locks? @batrick @gregsfortytwo What is dirlock? Seems to me like it is something existed and is got removed?
> 
> * https://github.com/ceph/ceph/blob/master/src/mds/MDCache.cc#L2105

I ran ```git log -G dirlock -- src/mds/```

and got

> commit fca6c27cbccd36d20d18a3592f13432821492617
> Author: Sage Weil <sweil@redhat.com>
> Date:   Thu Dec 18 10:36:47 2008 -0800
>
>     mds: kill the dirlock

LOL
*sigh*

So it looks like Locker::xlock_start() checks the current lock state and calls either simple_xlock() or simple_lock() based on that.
> if (lock->get_state() == LOCK_LOCK || lock->get_state() == LOCK_XLOCKDONE) {
  mut->start_locking(lock);
  simple_xlock(lock);
} else {
  simple_lock(lock);
}

And we ended up in the simple_lock branch and it aborted because it doesn't recognize the current state when we get there. I really don't know why that might have happened — it might be that we just need the wrlock, but it also might be that we need more state guards in place elsewhere whose semantics I don't have in my head any more.

Reply:
Yeh, that state isn't valid for scatterlock.

For the lock state machine in general, we would need extensive design write-up covering each state with its nuances to really make sense of what all states to add, which isn't going to be straightforward. In the interest of time switching to wrlock might _just_ be OK.